lib/Doctrine/ORM/Id/AbstractIdGenerator.php

```diff
@@ -20,22 +20,44 @@
 namespace Doctrine\ORM\Id;
 
 use Doctrine\ORM\EntityManager;
+use Doctrine\ORM\Id\IdGeneratorInterface;
+use Doctrine\ORM\ORMException;
+use Doctrine\ORM\EntityManagerInterface;
 
-abstract class AbstractIdGenerator
+abstract class AbstractIdGenerator implements IdGeneratorInterface
 {
     /**
-     * Generates an identifier for an entity.
+     * {@inheritDoc}
      *
      * @param EntityManager $em
      * @param \Doctrine\ORM\Mapping\Entity $entity
      * @return mixed
      */
     abstract public function generate(EntityManager $em, $entity);
 
+    public final function generateId(EntityManagerInterface $em, $entity)
+    {
+        /* @var $id mixed */
+        $id = null;
+
+        if ($em instanceof EntityManager) {
```

Comment:
Right, i forgot about exit early when writing this. My bad. I would switch the two cases around so that when more code get's added to the valid branch it would not be unnecessarily indented. That way the exception would act more like a guard clause. It would be easier to add more such clauses or any other code beneath it. Would you agree?
```
if (!$em instanceof EntityManager) {
    throw new ORMException(sprintf(
         "Tried to use non-doctrine entity-manager %s with old id-generator %s which is not supported! ".
         "This id-generator must be upgraded to use the %s to make it work with different entity-managers!",
         get_class($em),
         get_class($this),
         IdGeneratorInterface::class
    ));
}

return $this->generate($em, $entity);
```